conjure-codegen/src/context.rs

```diff
@@ -631,6 +631,95 @@ impl Context {
         }
     }
 
+    pub fn is_binary(&self, def: &Type) -> bool {
+        match def {
+            Type::Primitive(PrimitiveType::Binary) => true,
+            Type::Primitive(_)
+            | Type::Optional(_)
+            | Type::List(_)
+            | Type::Set(_)
+            | Type::Map(_) => false,
+            Type::Reference(def) => self.is_binary_ref(def),
+            Type::External(def) => self.is_binary(def.fallback()),
+        }
+    }
+
+    fn is_binary_ref(&self, name: &TypeName) -> bool {
+        let ctx = &self.types[name];
+
+        match &ctx.def {
+            TypeDefinition::Alias(def) => self.is_binary(def.alias()),
+            TypeDefinition::Enum(_) | TypeDefinition::Object(_) | TypeDefinition::Union(_) => false,
+        }
+    }
+
+    pub fn is_plain(&self, def: &Type) -> bool {
+        match def {
+            Type::Primitive(primitive) => match primitive {
+                PrimitiveType::String
+                | PrimitiveType::Datetime
+                | PrimitiveType::Integer
+                | PrimitiveType::Double
+                | PrimitiveType::Safelong
+                | PrimitiveType::Binary
+                | PrimitiveType::Boolean
+                | PrimitiveType::Uuid
+                | PrimitiveType::Rid
+                | PrimitiveType::Bearertoken => true,
+                PrimitiveType::Any => false,
+            },
+            Type::Optional(_) | Type::List(_) | Type::Set(_) | Type::Map(_) => false,
+            Type::Reference(def) => self.is_plain_ref(def),
+            Type::External(def) => self.is_binary(def.fallback()),
+        }
+    }
+
+    fn is_plain_ref(&self, name: &TypeName) -> bool {
+        let ctx = &self.types[name];
+
+        match &ctx.def {
+            TypeDefinition::Alias(def) => self.is_plain(def.alias()),
+            TypeDefinition::Enum(_) => true,
+            TypeDefinition::Object(_) | TypeDefinition::Union(_) => false,
+        }
+    }
+
+    pub fn is_single_value(&self, def: &Type) -> bool {
+        match def {
+            Type::Primitive(_) => true,
+            Type::Optional(_) | Type::List(_) | Type::Set(_) | Type::Map(_) => false,
+            Type::Reference(def) => self.is_single_value_ref(def),
+            Type::External(def) => self.is_single_value(def.fallback()),
+        }
+    }
+
+    fn is_single_value_ref(&self, name: &TypeName) -> bool {
+        let ctx = &self.types[name];
+
+        match &ctx.def {
+            TypeDefinition::Alias(def) => self.is_single_value(def.alias()),
+            TypeDefinition::Enum(_) | TypeDefinition::Object(_) | TypeDefinition::Union(_) => true,
```

Comment:
wait why are these true?

what is english definition of a 'single value'?

Reply:
Here, "single value" is in contrast to "iterable over multiple values". Maybe this should be inverted to `is_iterable`?

Reply:
might be clearer - alternatively just brief description of what this is (thought might be super obvious from context later on)